verbose output mode

kellyredding opened the pull request:
This adds a verbose runtime output mode flag to the CLI and config.
This also adds a method for querying this flag to the common view
helpers.  Finally, this updates the default view to honor the
flag in its output.

This goal here is to be able to give more details about what is
running at runtime.

Closes #208.

@jcredding ready for review

kellyredding commented:
Example output from assert's test suite:

```
$ assert --help
Usage: assert [options] [TESTS]

Options:
    -s, --runner-seed RUNNER-SEED    Use a given seed to run tests
    -o, --[no-]capture-output        capture stdout and display in result details
    -t, --[no-]halt-on-fail          halt a test when it fails
    -c, --[no-]changed-only          only run test files with changes
    -p, --[no-]pp-objects            pretty-print objects in fail messages
    -e, --[no-]profile               output test profile info
    -v, --[no-]verbose               output verbose runtime test info
    -d, --[no-]debug                 run in debug mode
        --version
        --help
$ assert -v
Loaded suite (439 tests)
Running tests in random order, seeded with "53265"
Assert::Factory::UnitTests: "Assert::Factory should return a random date using `date`"
  . 0.000674 seconds, 1 results, 1483.679525 results/s
Assert::Stub::ModuleTests: "Assert::Stub for singleton methods on a module should allow stubbing a method that takes a minimum number of args"
  .... 0.002049 seconds, 4 results, 1952.171791 results/s
Assert::Result::BaseTests: "Assert::Result::Base should respond to methods"
  ......... 0.000907 seconds, 9 results, 9922.822492 results/s
Assert::Context::SubjectDSL::DescriptionsTests: "Assert::Context::SubjectDSL `descriptions` method should return a collection containing any descriptions defined on the class"
  . 0.000303 seconds, 1 results, 3300.330033 results/s
Assert::Assertions::AssertFileExistsTests: "`assert_file_exists` should produce results as expected"
  ... 0.000522 seconds, 3 results, 5747.126437 results/s
Assert::Factory::UnitTests: "Assert::Factory should respond to methods"
  ................... 0.001281 seconds, 19 results, 14832.162373 results/s
Assert::Assertions::AssertNotEqualTests: "`assert_not_equal` should have a fail message with custom and generic explanations"
  . 0.000433 seconds, 1 results, 2309.468822 results/s
Assert::View::Base::UnitTests: "Assert::View::Base should respond to methods"
  ...................................... 0.002709 seconds, 38 results, 14027.316353 results/s
RunningSystemTests::PassTests: "Running a test (with no halt-on-fail) that passes a single assertion should have 1 result"
  . 0.000386 seconds, 1 results, 2590.673575 results/s

[...]
```

jcredding commented:
@kellyredding - Looks good :boom:

kellyredding commented:
@jcredding I'm thinking I should add the file and line number for each test to this output (after I merge #211).  What do you think?  Too noisy?  Where should it go?  On the line with the full test name and context class?  Replace the context class?

Would love your thoughts on this.  Thanks!

jcredding commented:
@kellyredding - It would be noisy but that's kinda the point of verbose mode. If you wanted to keep this as-is you could do the `-vv` which is like very verbose or something. I've seen that a few times. I'd be fine adding the path and line number to this though cause like I said, it's verbose.

As far as where, I'd probably put it on it's own line after the description and context class. I like the dots and the stats being the last line (mirrors the full test suite results in non-verbose mode) and the description and context class being the first (this is what I would be looking for to identify the test). I like the new line because if its a file path and line number its going to probably be longer and having a full line to read it seems easier. It's harder to read a file path If it gets wrapped.

kellyredding commented:
@jcredding good thoughts - thanks for the input.  That's kinda how I was leaning too.  I'll merge #211 and then update this.  Thanks man!

kellyredding commented:
@jcredding I've update this after rebasing off #211 and #213.  It now shows the file line info in the verbose output.  I've updated my comment above to reflect this.  I also added a README snippet for this option.  Please review the output tweaks and the README snippet.  Thanks!

jcredding commented:
@kellyredding - I like the indention, helps quickly find the test name. The readme snippet looks good and makes sense :boom:
